src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +30,162 @@
 
   private static final String FILTER_ARRAY = "clicked-array";
 
+  String[] allPlaces = {"Paris", "New York City", "San Francisco", "London", "Sydney", "Venice"};
+  String[] allDifficulties = {"Easy", "Medium", "Hard"};
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    HashSet<String> userPlaces = new HashSet();
+    HashSet<String> userDifficulties = new HashSet();
+
+    ArrayList<Destination> allDestinations = new ArrayList();
+    // ArrayList<Name of Destination>
+    ArrayList<String> filteredDestinations = new ArrayList();
+
     // Get array of clicked filters and convert to ArrayList<String>
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    allDestinations = createFakeDestinations();
+    userPlaces = separateFilters(clickedFilters, allPlaces);
+    userDifficulties = separateFilters(clickedFilters, allDifficulties);
+    filteredDestinations = filterPlaces(userPlaces, allDestinations);
+    filteredDestinations =
+        filterDifficulty(filteredDestinations, userDifficulties, allDestinations);
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
     response.getWriter().println(clickedFilters);
   }
+
+  public HashSet<String> separateFilters(HashSet<String> clickedFilters, String[] allTypeFilters) {
+    HashSet<String> userFilters = new HashSet();
+    for (int i = 0; i < allTypeFilters.length; i++) {
+      if (clickedFilters.contains(allTypeFilters[i])) {
+        userFilters.add(allTypeFilters[i]);
+      }
+    }
+    return userFilters;
+  }
+
+  /* Return Destination objects within specified place. */
+  public ArrayList<String> filterPlaces(
+      HashSet<String> userPlaces, ArrayList<Destination> allDestinations) {
+    ArrayList<String> filteredDestinations = new ArrayList();
+    // Iterate through destinations
+    // If Destination place is in userPlaces array
+    // Add Destination to filteredDestinations array
+    for (Destination currDestination : allDestinations) {
+      if (userPlaces.contains(currDestination.getCity())) {
+        filteredDestinations.add(currDestination.getName());
+      }
+    }
+    return filteredDestinations;
+  }
+
+  /* Return Destination objects with specified difficulty. */
+  public ArrayList<String> filterDifficulty(
+      ArrayList<String> filteredDestinations,
+      HashSet<String> userDifficulties,
+      ArrayList<Destination> allDestinations) {
```

Comment:
I'm a little confused by the complexity of this approach, because I think it would make sense to simply use streams again here.

filteredDestinations.stream().filter(destination -> userDifficulties .contains(destination. getDifficulty());

this requires some small changes to your existing code:
- filteredDestinations should be a list of _Destination_s and not names
-  userDifficulties should be a set of Destination.Obscurity levels that correspond to what the user is okay with.  If they are okay with everything it should be full of every level, not empty